ui/src/saros/ui/browser/SWTBrowser.java

```diff
@@ -0,0 +1,199 @@
+package saros.ui.browser;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.eclipse.swt.browser.Browser;
+import org.eclipse.swt.browser.BrowserFunction;
+import org.eclipse.swt.browser.ProgressEvent;
+import org.eclipse.swt.browser.ProgressListener;
+import org.eclipse.swt.events.DisposeEvent;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Display;
+
+public class SWTBrowser implements IBrowser {
+
+  private enum ReadyState {
+    LOADING,
+    COMPLETE
+  }
+
+  private final Browser browser;
+  private ReadyState state = ReadyState.LOADING;
+  private List<Runnable> runOnDisposalList =
+      Collections.synchronizedList(new ArrayList<Runnable>());
+
+  private final ProgressListener readyStateListener =
+      new ProgressListener() {
+        @Override
+        public void completed(ProgressEvent event) {
+          state = ReadyState.COMPLETE;
+        }
+
+        @Override
+        public void changed(ProgressEvent event) {
+          state = ReadyState.LOADING;
+        }
+      };
+
+  public SWTBrowser(Composite parent, int style) {
+    browser = new Browser(parent, style);
+
+    browser.addDisposeListener(
+        (DisposeEvent e) -> {
+          synchronized (runOnDisposalList) {
+            for (Runnable runnable : runOnDisposalList) {
+              runnable.run();
+            }
+          }
+        });
+  }
+
+  @Override
+  public boolean setFocus() {
+    return this.syncExec(
+        () -> {
+          return browser.setFocus();
+        });
+  }
+
+  @Override
+  public void setSize(final int width, final int height) {
+    this.syncExec(() -> browser.setSize(width, height));
+  }
+
+  @Override
+  public String getUrl() {
+    return this.syncExec(
+        () -> {
+          return browser.getUrl();
+        });
+  }
+
+  @Override
+  public boolean loadUrl(final String url, final int timeout) {
+    browser.getShell().open();
+
+    this.state = ReadyState.LOADING;
+    browser.addProgressListener(readyStateListener);
+
+    this.syncExec(
+        () -> {
+          browser.setUrl(url);
+        });
+    boolean successful = waitForPageLoaded(url, timeout);
+
+    browser.removeProgressListener(readyStateListener);
+
+    return successful;
+  }
+
+  private boolean waitForPageLoaded(final String url, final int timeout) {
+    AtomicBoolean waiting = new AtomicBoolean(true);
+    Display display = browser.getDisplay();
+
+    // Wake up the ui thread after the timeout exceeded
+    Runnable wakeUpTimeout =
+        () -> {
+          // avoid execution if url is already loaded
+          if (state != ReadyState.COMPLETE) {
+            waiting.set(false);
+            display.wake();
+          }
+        };
+
+    display.timerExec(timeout, wakeUpTimeout);
```

Comment:
The timeout related wake-up task and the event-loop are running on the ui thread. Therefore race-conditions between should be avoided. I know it is not perfect, because the `wakeUpTimeout` runnable is still executed (as NOP) even if the event-loop is already terminated. Other synchronization methods always allowed the case that `display.wake()` is executed after the check of the while loop, but before `display.sleep()`.
I tested this with our current html page (no timeout) and a html containing a `for (;;) {}` script (caused timeout).

Reply:
Just two questions.

Why must the class handle the UI thread access ? You are doing more work than neccessary.

If you call runAndDispatch you are able to call loadUrl again. I think it is better to abort with an illegalstate exception here because you start to load another html page.

Just create some async runnables that call loadUrl in these runables. They will not be executed one after another but all during the recursion.

See https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/ActivityHandler.java#L262

and assume modal context = readAndDispatch